### PR TITLE
feat(developer): add new touch layout special tags

### LINKED ISSUE #2738: ZWNJ image to put on a key
We need a ZWNJ to show to people on a key. It doesn't seem to be in unicode. But there is an ISO image shown here: https://en.wikipedia.org/wiki/Zero-width_non-joiner. You can see in the attached images what iOS and an Android keyboard does.
Draft
Persian Gboard Layout

### LINKED ISSUE #3863: feat(developer): add identifiers for new special touch layout keys
We need to add several new identifiers to the short list in Keyman Developer's touch layout editor; see #3851 for the codes and names.

> The relevant part for use within Developer.

_Originally posted by @jahorton in https://github.com/keymanapp/keyman/pull/3851#discussion_r521939995_

### PR DESCRIPTION
**Pull request by mcdurdin**
Fixes #3863.
Fixes #2738.

This adds the new `*RTLEnter*` and other tags introduced in #3851 to the touch layout editor, and adds validation for use in 14.0 keyboards.

Due to a limitation in the compiler architecture, use of these tags will not trigger the automatic keyboard versioning functionality; the keyboard developer must explicitly specify `store(&VERSION) '14.0'`.

Also:

* synchronised error codes between .h and .pas.
* synchronised touch layout tags between Developer and KeymanWeb, and cleaned up builder.js usage thereof.
* added support for KeymanwebOsk font override when special tags are identified.
* added basic tests to ensure that the compiler is catching use of these tags as needed.
* updated keymanweb-osk.ttf font in Developer.

### CODE REVIEW

**Review by jahorton — DISMISSED** [review]

**Comment by darcywong00** [comment]
If the .kmn sets 
```
store(&KMW_RTL) '1'
```
Will Developer automatically use the RTL versions in the touch layout, or will keyboard developers need to manually update their keyboards?

File: web/source/osk/visualKeyboard.ts [hunk 1]
```diff
@@ -537,6 +538,7 @@ namespace com.keyman.osk {
       '*RAltShift*':      0x68,
       '*LCtrlShift*':     0x69,
       '*RCtrlShift*':     0x70,
+      // Following codes introduced in Keyman 14.0
```
Comment (darcywong00) [inline web/source/osk/visualKeyboard.ts:541]: What about
```
      '*ShiftedLock*':    0x74,
      '*ZWNJ*':           0x75, // If this one is specified, auto-detection will kick in.
      '*ZWNJiOS*':        0x75, // The iOS version will be used by default, but the
      '*ZWNJAndroid*':    0x76
```
Are they going to be on a separate PR?

**Comment by darcywong00** [comment]
RE: ZWNJ
If someone defines a key U_200C, will they need to know to use the special "*ZWNJ*" key caps? Otherwise, won't they get a blank key cap?

**Comment by jahorton** [comment]
They'll need to implement some sort of key cap, yes.  `U_200C` is not automatically assigned one of these glyphs.

**Comment by mcdurdin** [comment]
Per discussion at Standup, we are considering a different approach to allow us to answer @darcywong00's suggestion around automatic key direction.

1. We'll add `*LTRBkSp*` and `*LTREnter*` special key labels which will always show LTR caps.
2. The `*BkSp*` and `*Enter*` labels in 14.0 onward will be bidi, using the `&KMW_RTL` value to determine directionality.
3. For down-version compatibility, we'll convert all 14.0 specials with a helper function compiled into the keyboard that runs on construction on the `KVKL` objects if the keyboard detects an older version of Keyman, as follows:

14.0 Special | 10.0 Conversion
------------|----------------
`*LTRBkSp*`, `*RTLBkSp*` | `*BkSp*`
`*LTREnter*`, `*RTLEnter*` | `*Enter*`
`*ShiftLock*` | `*Shift*`
`*ShiftedLock*` | `*Shifted*`
`*ZWNJ*`, `*ZWNJiOS*`, `*ZWNJAndroid*` | `>\|<` (i.e. just those 3 symbols)

**Comment by jahorton** [comment]
Fixes #2738.

**Comment by mcdurdin** [comment]
Note: all specials except for the ZWNJs will trigger a hint if they are used on a non-special key and the keyboard version is `< 14` -- this is not really supported (and not recommended in any case).

**Comment by mcdurdin** [comment]
This adds a new standard (internal) property to a compiled keyboard, `_v`, which at load time is set to the major version of Keyman that the keyboard is running under, with the following code:

`this._v=(typeof keyman!="undefined"&&typeof keyman.version=="string")?parseInt(keyman.version,10):9;`

This is then used in the KVKL to conditionalise the 14.0 supported specials with the following pattern:

`"text": this._v>13 ? "*ZWNJ*" : "<|>"`

File: windows/src/developer/TIKE/compile/CompileKeymanWeb.pas [hunk 2]
```diff
@@ -1641,19 +1655,17 @@ function TCompileKeymanWeb.ValidateLayoutFile(var sLayoutFile: string; const sVK
...
           ReportError(0, CWARN_TouchLayoutSpecialLabelOnNormalKey,
-            Format('Key "%s" on layout "%s", platform "%s" is not a Special key but has the special label "%s". This feature is only supported in Keyman 14 or later', [
-              FId, FLayer.Id, FPlatform.Name, FText
-            ]));
-        end
-        else if CSpecialText14.Contains(FText) and not IsKeyboardVersion14OrLater then
-        begin
-          ReportError(0, CWARN_TouchLayoutSpecialLabelRequires14,
-            Format('Key "%s" on layout "%s", platform "%s" has a special label "%s" which is only supported in Keyman 14 or later', [
+            Format('Key "%s" on layout "%s", platform "%s" does not have the key type "Special" or "Special (active)" but has the label "%s". This feature is only supported in Keyman 14 or later', [
```
Comment (jahorton) [inline windows/src/developer/TIKE/compile/CompileKeymanWeb.pas]: Breaking down the condition on a line-by-line basis would probably be clearer.  That conditional is pretty dense, even with the comment at the top, as the bit about "non-special keys in Keyman 13 or earlier" is extremely dense.  (It took me a few minutes to decompress the overall conditional in order to convince myself it was correct as a result.)

- `(CSpecialText10.Contains(FText) or CSpecialText14.Contains(FText))`:  if the key uses a `*special*` text label:
- `not (FKeyType in [tktSpecial, tktSpecialActive])`: keys of type 'special' have always had access to `*special*` labels
    - half of the comment's first line
- `not IsKeyboardVersion14OrLater`: all keys can use `*special*` labels in KMW 14.0+
    - the other half of the comment's first line
- `not CSpecialText14ZWNJ.Contains(FText)`:  if the `*special*` label corresponds to a ZWNJ, it is replaced with a non-`*special*` in versions of KMW before 14.0.

Alternatively, if might be clearer if the `not` lines were reworked a bit.  Instead of `not ____ and not ___`, it would be more easily parsed with a `not (____ or ____)` pattern.  Then, it would effectively read "make a warning on special keys _if_ used outside of these three valid cases".

To state the previous paragraph a bit differently, right now the `if` reads:  
- `if (2 conditions) and (1 condition) and (1 condition) and (1 condition)`
    - There are four top-level groups that seem to have equal weight in this pattern.  
- Alternatively:  `if (2 conditions) and (3 conditions)`
    - There are just two top-level groups, providing a hierarchical grouping that makes parsing the overall conditional easier.
    - The conditions within each group are more easily related to each other than to the conditionals of the other group.

File: windows/src/developer/TIKE/compile/CompileKeymanWeb.pas [hunk 3]
```diff
@@ -1950,6 +1982,8 @@ function TCompileKeymanWeb.WriteCompiledKeyboard: string; {UTF8}
     'function %s()%s'+
     '{%s'+
     '%s%s%s'+
+    // Following line caches the Keyman major version
+    '%sthis._v=(typeof keyman!="undefined"&&typeof keyman.version=="string")?parseInt(keyman.version,10):9;%s'+
```
Comment (jahorton) [inline windows/src/developer/TIKE/compile/CompileKeymanWeb.pas:1986]: Important note:  `this._v` is saved against the main keyboard object.

File: windows/src/developer/TIKE/compile/CompileKeymanWeb.pas [hunk 4]
```diff
@@ -1697,6 +1709,24 @@ function TCompileKeymanWeb.ValidateLayoutFile(var sLayoutFile: string; const sVK
...
+    begin
+      // Rewrite Special key labels that are only supported in Keyman 14+
+      // This code is a little ugly but effective.
+      if not IsKeyboardVersion14OrLater then
+      begin
+        for i := 0 to High(CSpecialText14Map) do
+        begin
+          // Assumes the JSON output format will not change
+          if FDebug
+            then sLayoutFile := ReplaceStr(sLayoutFile, '"text": "'+CSpecialText14Map[i][0]+'"', '"text": this._v>13 ? "'+CSpecialText14Map[i][0]+'" : "'+CSpecialText14Map[i][1]+'"')
```
Comment (jahorton) [inline windows/src/developer/TIKE/compile/CompileKeymanWeb.pas:1725]: If I'm not mistaken, this code would be run during construction of the `.KVKL` keyboard property - well, really, one of its sub-objects, right?  Would `this._v` then be referring to part of the `KVKL` object hierarchy under construction, or to the main keyboard-object's `_v` property?  `this`-scoping is well-known for being quite tricky in JS, so this feels like a possible weak point.

I'd like to see a freshly-compiled keyboard requiring use of the `_v` property tested within KMW to make sure the logic holds correctly.  While screenshots as proof would probably suffice, it'd be nice to have the actual compiled keyboard (used for those screenshots) for independent test use as well.

**Comment by jahorton** [comment]
Oh, one more tidbit:  I believe we still need to define `*LTREnter*` and `*LTRBkSp*` within KMW; those haven't yet been added.

What has been added so far: https://github.com/keymanapp/keyman/blob/0b620957262976c4b580a5cbdf0e24023648ce24/web/source/osk/visualKeyboard.ts#L541-L548

File: windows/src/developer/TIKE/compile/CompileKeymanWeb.pas (hunk 2 above)
Comment (mcdurdin) [reply windows/src/developer/TIKE/compile/CompileKeymanWeb.pas]: I'm not sure that makes the conditional any easier to read, for me: I find negating groups of conditions is harder to map mentally than the current, reasonably flat list of conditions. Think of the first pair of conditions as `[].concat(CSpecialText10, CSpecialText14).includes(foo)` (though that would be much more unwieldy in Delphi code)?

But it's a complex set of conditions, and no matter which way you construct it, it's going to take a bit of mental parsing.

File: windows/src/developer/TIKE/compile/CompileKeymanWeb.pas (hunk 3 above)
Comment (mcdurdin) [reply windows/src/developer/TIKE/compile/CompileKeymanWeb.pas:1986]: Yes, and I am intending it for internal-use only (given it is a cache of a KeymanWeb property anyway)

File: windows/src/developer/TIKE/compile/CompileKeymanWeb.pas (hunk 4 above)
Comment (mcdurdin) [reply windows/src/developer/TIKE/compile/CompileKeymanWeb.pas:1725]: > this code would be run during construction of the `.KVKL` keyboard property

Not quite. It's an assignment of an object literal, so has the scope of the containing object.

> I'd like to see a freshly-compiled keyboard requiring use of the `_v` property tested within KMW to make sure the logic holds correctly.

`test_touchlayout_14_3.kmn` can be compiled to demonstrate the solution. (I can't currently attach files to the issue due to a GH glitch, so here is a [zip](https://drive.google.com/file/d/10TiSrdl5jBjBVcXjLpNyy0l7Kdn873O5/view?usp=sharing) on Google Drive.)

I have manually tested within KeymanWeb and encountered no issues.

File: windows/src/developer/TIKE/compile/CompileKeymanWeb.pas [hunk 5]
```diff
@@ -1625,6 +1649,27 @@ function TCompileKeymanWeb.ValidateLayoutFile(var sLayoutFile: string; const sVK
...
+
+      //
+      // Check that if the key has a *special* label, it is available in the target version
+      //
+      if FText.StartsWith('*') and FText.EndsWith('*') and (FText.Length > 2) then
+      begin
+        // We do not support '*special*' labels on non-special keys in Keyman 13
+        // or earlier, except for ZWNJ, which will be transformed in function
+        // TransformSpecialKeys14 to '<|>', and  which does not require the custom
+        // OSK font.
```
Comment (jahorton) [inline windows/src/developer/TIKE/compile/CompileKeymanWeb.pas:1661]: Just a suggestion, as this version of the comment feels clearer to me.  Feel free to disregard it.

```suggestion
        // Keyman versions before 14 do not support '*special*' labels on non-special keys.
        // ZWNJ use, however, is safe because it will be transformed in function
        // TransformSpecialKeys14 to '<|>',  which does not require the custom OSK font.
```

**Review by jahorton — APPROVED** [review]

**Comment by rmlockwood** [comment]
I really appreciate the new RTL images. What about the zero-width joiner, though? Can we get an image for it as well?

**Comment by mcdurdin** [comment]
@rmlockwood, yes, that's a good point -- not sure how we missed that. There may be others to consider as well -- maybe ZWSP? Do you have any suggestions on what ZWJ and ZWSP could look like?

**Comment by rmlockwood** [comment]
@mcdurdin, ZWSP isn’t relevant for our Arabic script languages so no suggestions. RTL & LTR – right to left marker and left to right marker are somewhat relevant, more on desktop keyboards. I don’t have image suggestions for these. As far as ZWJ, I haven’t found any keyboards on iOS and Android that show it. My only suggestion would be to used a similar image to the ZWNJ on Android, just with the arrows pointing in. Maybe even use this image on iOS.
